src/backend/distributed/commands/multi_copy.c

```diff
@@ -218,19 +403,61 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
 		}
 	}
 
-	partitionMethod = PartitionMethod(tableId);
-	if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod == DISTRIBUTE_BY_RANGE)
-	{
-		CopyToExistingShards(copyStatement, completionTag);
-	}
-	else if (partitionMethod == DISTRIBUTE_BY_APPEND)
+	workerCopy = WorkerCopy(copyStatement);
+	if (workerCopy)
 	{
-		CopyToNewShards(copyStatement, completionTag);
+		NodeAddress *masterNodeAddress = MasterNodeAddress(copyStatement);
+		char *nodeName = masterNodeAddress->nodeName;
+		int32 nodePort = masterNodeAddress->nodePort;
+		PGconn *masterConnection = ConnectToNode(nodeName, nodePort);
+
+		char *relationName = copyStatement->relation->relname;
+		PGresult *queryResult = NULL;
+
+		char partitionMethod = RemotePartitionMethod(masterConnection, relationName);
+		if (partitionMethod != DISTRIBUTE_BY_APPEND)
+		{
+			ereport(ERROR, (errmsg("copy from worker nodes are only supported "
```

Comment:
If something fails, we should close the master connection. Otherwise, it remains open for the duration of the session. Alternatively, you could use GetOrEstablishConnection to cache the connection. Not sure if there is any value in caching it, but at least you wouldn't open a new connection every time.